Android datepicker mode configurations added

pandiaraj44 opened the pull request:
# **Motivation**
Currently, The **DatePickerAndroid** component opens the native date picker with default mode. We can't able to change the mode via configurations. To support android **[date-picker mode](https://developer.android.com/reference/android/widget/DatePicker.html)**, existing component needs to be changed.

# Solution
**For Android >= 5.0**, The DatePickerDialog doesn't provide the default method to change the mode of a date picker.  So I have added custom theme which will support two kinds of **mode('spinner','calendar')** ref:https://developer.android.com/reference/android/R.attr.html#datePickerStyle.

**For Android < 5.0,** The DatePickerDialog provides the default method to change the mode of a date picker. ref:https://developer.android.com/reference/android/widget/DatePicker.html#setCalendarViewShown(boolean).

With the help of **Build.VERSION.SDK_INT** I have done the above functionality with limited lines of code changes and also I have added the example to UIExplorer.

# **Test Plan**
When opening a date picker on Android, We need to pass the option value(mode: 'calendar/spinner/default'). 

## **Example1:** 
DatePickerAndroid.open({ date: Date.now(), mode: 'calendar'}) // It will open datepicker with calendar mode

**Android < 5.0**
![android_j 5 0 _calendar](https://cloud.githubusercontent.com/assets/22169327/20267264/e0333a70-aa9f-11e6-9c76-8057f7b5dbad.png)

**Android >= 5.0**
![android_m 5 0 _calendar](https://cloud.githubusercontent.com/assets/22169327/20267265/e0338cfa-aa9f-11e6-8d90-503ef0c8de6f.png)

## 
**Example2:** 
DatePickerAndroid.open({ date: Date.now(), mode: 'spinner'}) // It will open datepicker with spinner mode

**Android < 5.0**
![android_j 5 0 _spinner](https://cloud.githubusercontent.com/assets/22169327/20267319/24ea9f50-aaa0-11e6-9ab2-483caff32bc0.png)

**Android >= 5.0**
![android_m 5 0 _spinner](https://cloud.githubusercontent.com/assets/22169327/20267326/2e1bc270-aaa0-11e6-8086-0b93e65efcb4.png)


## **Example3:** 
DatePickerAndroid.open({ date: Date.now()})
(or)
DatePickerAndroid.open({ date: Date.now(), mode: 'default'})
 // It will open datepicker with default native mode based on android versions.

**Android < 5.0**
![android_j 5 0 _spinner](https://cloud.githubusercontent.com/assets/22169327/20267319/24ea9f50-aaa0-11e6-9ab2-483caff32bc0.png)

**Android >= 5.0**
![android_m 5 0 _calendar](https://cloud.githubusercontent.com/assets/22169327/20267265/e0338cfa-aa9f-11e6-8d90-503ef0c8de6f.png)

So If the user doesn't give mode option then default mode will be shown.

For real-time testing, you can run UIExplorer android project. I have added example code.
![screenshot_20161119-123805](https://cloud.githubusercontent.com/assets/22169327/20453779/6395e878-ae55-11e6-9838-704b32be3bd6.png)

ericvicenti commented:
This looks reasonable. Why is it tagged "[URGENT]"? Can you verify that the test failures are unrelated to this change?

pandiaraj44 commented:
@ericvicenti removed the urgent tag from title. The build issues are related to Xcode8, I have changed only the DatePickerDialogModule for android.

ericvicenti commented:
Ok, looks great!

@facebook-github-bot shipit

ericvicenti commented:
We had some test failures internally. Do you know why this would happen?

```
/data/react-native-github/ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:28: error: cannot find symbol
import com.facebook.react.R;
                         ^
/data/react-native-github/ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:63: error: package R does not exist
        dialog = new DismissableDatePickerDialog(activityContext, R.style.CalendarDatePickerDialog, onDateSetListener, year, month, day);
                                                                   ^
/data/react-native-github/ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:65: error: package R does not exist
        dialog = new DismissableDatePickerDialog(activityContext, R.style.SpinnerDatePickerDialog, onDateSetListener, year, month, day);
stderr: /data/react-native-github/ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:28: error: cannot find symbol
```

pandiaraj44 commented:
@ericvicenti I have added custom styles in "react-native/ReactAndroid/src/main/res/devsupport/values/styles.xml". For reflecting those styles you have to run a clean build on android projects.

ericvicenti commented:
Hmm. Maybe @mkonicek can help land this, because I have no idea what I'm doing when it comes to best practices around these things.

pandiaraj44 commented:
@ericvicenti , @mkonicek I have run the "npm test" command inside react-native folder. The test cases are executed successfully. 
FYI...
![screenshot from 2016-11-15 14 38 20](https://cloud.githubusercontent.com/assets/22169327/20299620/c26f8d8e-ab41-11e6-928f-311c4baa54c7.png)
 
Can someone from android core team have a look at on this.

leeight commented:
Try to run `scripts/run-android-local-unit-tests.sh`

pandiaraj44 commented:
Hi @ericvicenti , @ericvicenti , @facebook-github-bot , @leeight Finally, I have found the root cause of the issue("error: cannot find symbol import com.facebook.react.R;"). I have missed the resource dependency in date picker buck file. Now I have added the dependencies to  "ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/BUCK"  file. I run the android unit test cases and it was passed and In an android integration test,  date picker modal test cases were passed.

FYI..
sh scripts/run-android-local-unit-tests.sh- Result
![screenshot from 2016-11-16 12 12 40](https://cloud.githubusercontent.com/assets/22169327/20337936/05b6231a-abfa-11e6-816b-c3a512fdedae.png)

sh scripts/run-android-local-integration-tests.sh - Result
![screenshot from 2016-11-16 12 12 53](https://cloud.githubusercontent.com/assets/22169327/20337958/30175c5a-abfa-11e6-846e-ad2f4ba34b1d.png)

I have re-base the code from master and updated the pull request.

pandiaraj44 commented:
@AaaChiuuu / @spicyj Can you please review the pull request.

spicyj commented:
@pandiaraj44 Please stop tagging everyone. It sounds like @mkonicek will look into why the import failed and tell you if anything needs to be changed.

ericvicenti commented:
I am re-running internal tests and will land this afternoon if everything passes.

leeight commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/BUCK:
```diff
@@ -10,6 +10,7 @@ android_library(
     react_native_target('java/com/facebook/react/bridge:bridge'),
     react_native_target('java/com/facebook/react/common:common'),
     react_native_target('java/com/facebook/react/module/annotations:annotations'),
+    react_native_target('res:shell'),
```
Actually i'm not sure it's right or not add  `res:shell` dependency here, but i guess it's a bad smell...

pandiaraj44 replied:
Inside ReactAndroid project, The shell resource visibility is in PUBLIC mode. So I have reused that one and added some list of styles to shell resource folder. 

I don't think it is necessary to create a resource folder of each of the package in ReactAndroid project. 

Maybe if the visibility of the shell is in private mode then I would create separate resource folder.

Ref: https://developer.android.com/guide/topics/ui/themes.html

pandiaraj44 commented on Libraries/Components/DatePickerAndroid/DatePickerAndroid.android.js:
```diff
@@ -52,14 +52,15 @@ class DatePickerAndroid {
    *   * `date` (`Date` object or timestamp in milliseconds) - date to show by default
    *   * `minDate` (`Date` or timestamp in milliseconds) - minimum date that can be selected
    *   * `maxDate` (`Date` object or timestamp in milliseconds) - minimum date that can be selected
+   *   * `calendarMode` (`Boolean` true or false) - To set the date-picker mode to spinner/calendar.
```
As per the android API documentation, Android date-picker supports only two modes. https://developer.android.com/reference/android/R.styleable.html#DatePicker_datePickerMode

pandiaraj44 commented:
Hi @ericvicenti , FYI... All git hub checks have passed.

pandiaraj44 commented:
Hi @ericvicenti , Again I have re-base my code from upstream and tests have passed.

satya164 commented:
I think an enum is a better API for this, even if it's just two types, `mode='calender'` and `mode='spinner'` feel better to me than `calenderMode`.

mkonicek commented:
Thanks for doing this! I have one API request: rather than using boolean, how about being explicit:

    DatePickerAndroid.open({ date: Date.now(), mode: 'calendar'})
    DatePickerAndroid.open({ date: Date.now(), mode: 'spinner'})

Rather than `calendarMode: false` the reader can immediately see this is using the spinners. It also allows for adding a third option in the future without introducing a breaking API change.

pandiaraj44 commented:
Thanks @leeight , @satya164 , @mkonicek. It is a great idea we can go with the enum. I have added another mode.

enum('calendar', 'spinner', 'default')

-  'calendar': Show a date picker in calendar mode.
-  'spinner': Show a date picker in spinner mode.
-  'default': Show a default native date picker(spinner/calendar) based on android versions.

pandiaraj44 commented:
Hi @leeight , I think we can go with getResources().getIdentifier(). I have removed shell dependency.

```
dialog = new DismissableDatePickerDialog(activityContext,
            activityContext.getResources().getIdentifier("CalendarDatePickerDialog", "style", activityContext.getPackageName()),
            onDateSetListener, year, month, day);
```

leeight commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,48 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
-    final DatePickerDialog dialog =
-        new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);
+    DatePickerMode mode = DatePickerMode.DEFAULT;  //To set mode to default
+    if(args != null && args.getString(DatePickerDialogModule.ARG_MODE, null) != null) {
```
`if (`

leeight commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,48 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+
+    DatePickerDialog dialog = null;
+
+    if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) { // To support material calendar date-picker
```
`if (`

pandiaraj44 replied:
I have updated the pull request.

leeight commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,48 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+          dialog = new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);
+          break;
+      }
+    }else {
```
`} else {`

pandiaraj44 replied:
I have updated the pull request.

leeight commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,48 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+          //To Show a default native date picker
+          dialog = new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);
+          break;
+      }
```
```java
dialog = new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);

switch (mode) {
  case CALENDAR:
    dialog.getDatePicker().setCalendarViewShown(true);
    dialog.getDatePicker().setSpinnersShown(false);
    break;
  case SPINNER:
    dialog.getDatePicker().setCalendarViewShown(false);
    break;
}
```

pandiaraj44 replied:
I have updated the pull request. Thanks @leeight

pandiaraj44 commented:
Hi @mkonicek , Changes are done. Can you review it.

pandiaraj44 commented:
cc @leeight

pandiaraj44 commented:
cc @satya164 @ericvicenti @facebook-github-bot

satya164 commented:
@pandiaraj44 Can you address @leeight 's nits regarding whitespace please?

pandiaraj44 commented:
@satya164 Updated the pull request.
Can you please let me know who is the potential reviewer for this pull request and when they will accept/reject this pull request?

cc @leeight

satya164 commented:
> Can you please let me know who is the potential reviewer

Me, @leeight and @mkonicek are reviewing your PR, so I guess we're the reviewers

> when they will accept/reject this pull request?

We'll merge when all review comments have been addressed

pandiaraj44 commented on ReactAndroid/src/androidTest/assets/AndroidTestBundle.js.meta:
```diff
@@ -0,0 +1 @@
+5	 4f:}&�		dl�8mH
```
@satya164 While running test cases that file get created. I removed that file and updated the pull request.

pandiaraj44 commented:
@satya164 Thank you very much for your information.

mkonicek commented:
@pandiaraj44 Please comment once you've addressed all the comments.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,44 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
 
-    final DatePickerDialog dialog =
-        new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);
+    DatePickerMode mode = DatePickerMode.DEFAULT;  //To set mode to default
```
nit: Please remove comment that repeats what the code does without adding information.

pandiaraj44 replied:
Comment has been removed.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,44 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+            onDateSetListener, year, month, day);
+          break;
+        case DEFAULT:
+          //To Show a default native date picker
```
Please remove redundant comment.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,44 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+          break;
+        case SPINNER:
+          dialog.getDatePicker().setCalendarViewShown(false);
+          break;
```
The code in this switch statement should be moved up to the branches where the dialogs are created.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerDialogFragment.java:
```diff
@@ -53,8 +54,44 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
+          break;
+      }
+    } else {
+      dialog = new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);
```
It's possible to write the code above without repeating this line twice:

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      if (mode ==  DatePickerMode.CALENDAR) {
        // Create picker with calendar
      } else (if mode == DatePickerMode.SPINNER) {
        // Create picker with spinner
      }
    }
    // Default case

pandiaraj44 replied:
setCalendarViewShown has been deprecated in latest android API version. So build version **"if condition**" is mandatory. 

We can remove the dialog creation line but it requires extra two if condition
```
    DatePickerDialog dialog = null;

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      switch (mode) {
        case CALENDAR:
          dialog = new DismissableDatePickerDialog(activityContext,
            activityContext.getResources().getIdentifier("CalendarDatePickerDialog", "style", activityContext.getPackageName()),
            onDateSetListener, year, month, day);
          break;
        case SPINNER:
          dialog = new DismissableDatePickerDialog(activityContext,
            activityContext.getResources().getIdentifier("SpinnerDatePickerDialog", "style", activityContext.getPackageName()),
            onDateSetListener, year, month, day);
          break;
      }
    }

    if (dialog == null) {
      dialog = new DismissableDatePickerDialog(activityContext, onDateSetListener, year, month, day);
    }

    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      switch (mode) {
        case CALENDAR:
          dialog.getDatePicker().setCalendarViewShown(true);
          dialog.getDatePicker().setSpinnersShown(false);
          break;
        case SPINNER:
          dialog.getDatePicker().setCalendarViewShown(false);
          break;
      }
    }


```

For supporting future cases, I have used a switch statement.

mkonicek replied:
Makes sense, thanks!

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/datepicker/DatePickerMode.java:
```diff
@@ -0,0 +1,39 @@
+    mValue = value;
+  }
+
+  public static DatePickerMode fromString(String value) {
```
Would `valueOf` work instead?
http://stackoverflow.com/questions/604424/lookup-enum-by-string-value

pandiaraj44 replied:
ValueOf has been implemented. For that, we need to match the user values with enum constants.

mkonicek commented on ReactAndroid/src/main/res/shell/values/styles.xml:
```diff
@@ -10,4 +10,22 @@
+
+    <style name="CalendarDatePickerStyle" parent="android:Widget.Material.Light.DatePicker" tools:targetApi="lollipop">
+        <item name="android:datePickerMode">calendar</item>
+    </style>
```
Is it needed for these styles to be defined in resources? Is it possible to set the `android:datePickerMode` programmatically instead?

pandiaraj44 replied:
Latest Android API doesn't provide the default methods to change the mode.
Ref: https://developer.android.com/reference/android/widget/DatePicker.html#attr_android:calendarViewShown ,
https://developer.android.com/reference/android/widget/DatePicker.html#attr_android:datePickerMode

mkonicek commented:
Thanks!

@facebook-github-bot shipit
